File: ios/RNAliMediaDownloader.m
//
//  RNAliMediaDownloader.m
//  AliyunVideoClient_Entrance
//
//  Created by wsk on 2022/12/6.
//  Copyright © 2022 Aliyun. All rights reserved.
//

#import "RNAliMediaDownloader.h"
#import <AliyunMediaDownloader/AliyunMediaDownloader.h>
#import <React/RCTLog.h>
#import "AliMediaDownloader+vid.h"

NSString* const EK_downloaderOnPrepared             = @"downloaderOnPrepared";
NSString* const EK_downloaderOnError                = @"downloaderOnError";
NSString* const EK_downloaderOnDownloadingProgress  = @"downloaderOnDownloadingProgress";
NSString* const EK_downloaderOnProcessingProgress   = @"downloaderOnProcessingProgress";
NSString* const EK_downloaderOnCompletion           = @"downloaderOnCompletion";

@interface RNAliMediaDownloader ()
@property (nonatomic,strong)NSMutableDictionary<NSString*, AliMediaDownloader*> *preDownloaders;
@property (nonatomic,strong)NSMutableDictionary<NSString*, AliMediaDownloader*> *downloaders;
@end

@implementation RNAliMediaDownloader

RCT_EXPORT_MODULE(RNAliMediaDownloader)

/**
 @brief 初始化下载对象
 */
- (instancetype)init {
  if (self = [super init]) {
  }
  return self;
}

-(NSArray<NSString *> *)supportedEvents{
    return @[
      EK_downloaderOnPrepared,
      EK_downloaderOnError,
      EK_downloaderOnDownloadingProgress,
      EK_downloaderOnProcessingProgress,
      EK_downloaderOnCompletion
    ];
}

- (NSDictionary *)constantsToExport {
    return @{
        @"onPrepared": EK_downloaderOnPrepared,
        @"onError": EK_downloaderOnError,
        @"onProcessingProgress": EK_downloaderOnProcessingProgress,
        @"onDownloadingProgress": EK_downloaderOnDownloadingProgress,
        @"onCompletion": EK_downloaderOnCompletion
    };
}

- (NSString*)mapKey:(NSString*)vid trackIndex:(int)trackIndex {
  NSString *key = [NSString stringWithFormat:@"%@-%@", vid, @(trackIndex)];
  return key;
}

/**
 @brief 删除下载文件
 @param saveDir 文件保存路径
 @param vid     vid
 @param format  格式
 @param index   vid对应的下载索引
 */
RCT_EXPORT_METHOD(deleteFile:(NSString*)saveDir
                  vid:(NSString*)vid
                  format:(NSString*)format
                  index:(int)index
                  callback:(RCTResponseSenderBlock)callback) {
    RCTLogInfo(@"rn_media_downloader deleteFile vid = %@ trackIndex = %d", vid, index);

    int result = [AliMediaDownloader deleteFile:saveDir vid:vid format:format index:index];
    callback(@[[NSNumber numberWithInt:result]]);
    
    /// 如果是准备中、下载中删除
    NSString *key = [self mapKey:vid trackIndex:index];
    AliMediaDownloader *downloader = [self.downloaders objectForKey:key];
    if (downloader) {
      RCTLogInfo(@"rn_media_downloader deleteFile downloaders");

      [downloader deleteFile];
      [downloader destroy];
      self.downloaders[key] = nil;
    }

    downloader = [self.preDownloaders objectForKey:vid];
    if (downloader) {
      RCTLogInfo(@"rn_media_downloader deleteFile preDownloaders");
      [downloader destroy];
      self.preDownloaders[vid] = nil;
    }
}

/**
 @brief 销毁下载对象
 */
RCT_EXPORT_METHOD(destroyWithVid:(NSString*)vid trackIndex:(int)trackIndex) {
    
  RCTLogInfo(@"rn_media_downloader destroyWithVid vid = %@ trackIndex = %d", vid, trackIndex);

  NSString *key = [self mapKey:vid trackIndex:trackIndex];
  
  AliMediaDownloader *downloader = [self.downloaders objectForKey:key];
  if (downloader) {
    [downloader destroy];
    self.downloaders[key] = nil;
  }

  downloader = [self.preDownloaders objectForKey:vid];
  if (downloader) {
    [downloader destroy];
    self.preDownloaders[vid] = nil;
  }
}

/**
 @brief 使用AVPVidStsSource准备播放
 @param source vid sts的播放方式
 */
RCT_EXPORT_METHOD(prepareWithVid:(NSString*)vid
                  accessKeyId:(NSString *) accessKeyId
                  accessKeySecret:(NSString *) accessKeySecret
                  securityToken:(NSString *) securityToken
                  region:(NSString *) region) {
    RCTLogInfo(@"rn_media_downloader prepareWithVid vid = %@", vid);

    AVPVidStsSource *source = [[AVPVidStsSource alloc] initWithVid:vid
                                                      accessKeyId:accessKeyId
                                                  accessKeySecret:accessKeySecret
                                                    securityToken:securityToken
                                                            region:region];
  
    AliMediaDownloader * downloader = [[AliMediaDownloader alloc] init];
    downloader.mm_vid = vid;
    downloader.delegate = (id<AMDDelegate>)self;
    self.preDownloaders[downloader.mm_vid] = downloader;
    [downloader prepareWithVid:source];
}

/**
 @brief 鉴权过期，更新AVPVidStsSource信息，
 @param source vid sts的信息
 */
RCT_EXPORT_METHOD(updateWithVid:(NSString*)vid
                  trackIndex:(int) trackIndex
                  accessKeyId:(NSString *) accessKeyId
                  accessKeySecret:(NSString *) accessKeySecret
                  securityToken:(NSString *) securityToken
                  region:(NSString *) region) {
    RCTLogInfo(@"rn_media_downloader updateWithVid vid = %@ trackIndex = %d", vid, trackIndex);

    AVPVidStsSource *source = [[AVPVidStsSource alloc] initWithVid:vid
                                                      accessKeyId:accessKeyId
                                                  accessKeySecret:accessKeySecret
                                                    securityToken:securityToken
                                                            region:region];
    NSString *key = [self mapKey:vid trackIndex:trackIndex];
    AliMediaDownloader *downloader = [self.downloaders objectForKey:key];
    [downloader updateWithVid:source];
}

/**
 @brief 设置下载的保存路径
 @param dir 保存文件夹
 */
RCT_EXPORT_METHOD(setSaveDirectory:(NSString*)dir vid:(NSString*)vid trackIndex:(int)trackIndex) {
    RCTLogInfo(@"rn_media_downloader setSaveDirectory vid = %@ trackIndex = %d", vid, trackIndex);

    NSString *key = [self mapKey:vid trackIndex:trackIndex];
    AliMediaDownloader *downloader = [self.downloaders objectForKey:key];
    [downloader setSaveDirectory:dir];
}

/**
 @brief 开始下载
 */
RCT_EXPORT_METHOD(startWithVid:(NSString*)vid trackIndex:(int)trackIndex) {
    RCTLogInfo(@"rn_media_downloader startWithVid vid = %@ trackIndex = %d", vid, trackIndex);

    NSString *key = [self mapKey:vid trackIndex:trackIndex];
    AliMediaDownloader *downloader = [self.downloaders objectForKey:key];
    [downloader start];
}

/**
 @brief 停止下载
 */
RCT_EXPORT_METHOD(stopWithVid:(NSString*)vid trackIndex:(int)trackIndex) {
    RCTLogInfo(@"rn_media_downloader stopWithVid vid = %@ trackIndex = %d", vid, trackIndex);
    
    NSString *key = [self mapKey:vid trackIndex:trackIndex];
    AliMediaDownloader *downloader = [self.downloaders objectForKey:key];
    [downloader stop];
}

/**
 @brief 获取下载config
 */
RCT_EXPORT_METHOD(getConfigWithVid:(NSString*)vid trackIndex:(int)trackIndex callback:(RCTPromiseResolveBlock)callback) {
    RCTLogInfo(@"rn_media_downloader getConfigWithVid vid = %@ trackIndex = %d", vid, trackIndex);

    NSString *key = [self mapKey:vid trackIndex:trackIndex];
    AliMediaDownloader *downloader = [self.downloaders objectForKey:key];
    AVDConfig *config = [downloader getConfig];
    NSMutableDictionary *dic = [NSMutableDictionary dictionary];
    dic[@"timeoutMs"] = @(config.timeoutMs);
    dic[@"connnectTimoutMs"] = @(config.connnectTimoutMs);
    dic[@"referer"] = config.referer;
    dic[@"userAgent"] = config.userAgent;
    dic[@"httpProxy"] = config.httpProxy;
    callback(dic);
}

/**
 @brief 设置下载config
 */
RCT_EXPORT_METHOD(setConfig:(int)timeoutMs
                  connnectTimoutMs:(int)connnectTimoutMs
                  referer:(NSString*)referer
                  userAgent:(NSString*)userAgent
                  httpProxy:(NSString*)httpProxy
                  vid:(NSString*)vid
                  trackIndex:(int)trackIndex
) {
    RCTLogInfo(@"rn_media_downloader setConfig vid = %@ trackIndex = %d", vid, trackIndex);

    AVDConfig *config = [[AVDConfig alloc] init];
    config.timeoutMs = timeoutMs;
    config.connnectTimoutMs = connnectTimoutMs;
    config.referer = referer;
    config.userAgent = userAgent;
    config.httpProxy = httpProxy;
  
    NSString *key = [self mapKey:vid trackIndex:trackIndex];
    AliMediaDownloader *downloader = [self.downloaders objectForKey:key];
    [downloader setConfig:config];
}

/**
 @brief 设置下载的trackIndex
 @param trackIndex 从prepare回调中可以获取所有index
 */
RCT_EXPORT_METHOD(selectTrack:(int)trackIndex vid:(NSString*)vid) {
    RCTLogInfo(@"rn_media_downloader selectTrack vid = %@ trackIndex = %d", vid, trackIndex);

    AliMediaDownloader *downloader = self.preDownloaders[vid];

    downloader.mm_trackIndex = @(trackIndex).stringValue;
    [downloader selectTrack:trackIndex];
  
    NSString *key = [self mapKey:vid trackIndex:trackIndex];
    self.downloaders[key] = downloader;
  
    self.preDownloaders[vid] = nil;
}

/**
 @brief 获取SDK版本号信息
 */
RCT_EXPORT_METHOD(getSDKVersion:(RCTPromiseResolveBlock)callback) {
    RCTLogInfo(@"rn_media_downloader getSDKVersion");

    NSString *version = [AliMediaDownloader getSDKVersion];
    callback(version);
}

- (NSMutableDictionary<NSString*,AliMediaDownloader*> *)downloaders {
  if(!_downloaders) {
    _downloaders = [[NSMutableDictionary<NSString*,AliMediaDownloader*> alloc] init];
  }
  return _downloaders;
}

- (NSMutableDictionary<NSString*,AliMediaDownloader*> *)preDownloaders {
  if(!_preDownloaders) {
    _preDownloaders = [[NSMutableDictionary<NSString*,AliMediaDownloader*> alloc] init];
  }
  return _preDownloaders;
}
@end

#pragma mark - <AMDDelegate>
@interface RNAliMediaDownloader (AMDDelegate)<AMDDelegate>
@end
@implementation RNAliMediaDownloader (AMDDelegate)
/**
 @brief 下载准备完成事件回调
 @param downloader 下载downloader指针
 @param info 下载准备完成回调，@see AVPMediaInfo
 */
-(void)onPrepared:(AliMediaDownloader*)downloader mediaInfo:(AVPMediaInfo*)info {
  RCTLogInfo(@"rn_media_downloader onPrepared vid = %@", downloader.mm_vid);

  NSMutableArray *tracks = [NSMutableArray array];
  
  [info.tracks enumerateObjectsUsingBlock:^(AVPTrackInfo * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
    NSMutableDictionary *track = [NSMutableDictionary dictionary];
    track[@"trackIndex"] = @(obj.trackIndex);
    track[@"trackDefinition"] = obj.trackDefinition;
    track[@"vodFormat"] = obj.vodFormat;
    track[@"vodFileSize"] = @(obj.vodFileSize);
    [tracks addObject:track];
  }];
  
  NSMutableDictionary *body = [NSMutableDictionary dictionary];
  body[@"vid"] = downloader.mm_vid;
  body[@"tracks"] = tracks;
  [self sendEventWithName:EK_downloaderOnPrepared body:body];
}

/**
 @brief 错误代理回调
 @param downloader 下载downloader指针
 @param errorModel 播放器错误描述，参考AliVcPlayerErrorModel
 */
- (void)onError:(AliMediaDownloader*)downloader errorModel:(AVPErrorModel *)errorModel {
  RCTLogInfo(@"rn_media_downloader onError vid = %@ code = %@", downloader.mm_vid, @(errorModel.code));

  NSMutableDictionary *body = [NSMutableDictionary dictionary];
  body[@"vid"] = downloader.mm_vid;
  body[@"trackIndex"] = downloader.mm_trackIndex;
  body[@"code"] = @(errorModel.code);
  body[@"message"] = errorModel.message;
  [self sendEventWithName:EK_downloaderOnError body:body];
}

/**
 @brief 下载进度回调
 @param downloader 下载downloader指针
 @param percent 下载进度 0-100
 */
- (void)onDownloadingProgress:(AliMediaDownloader*)downloader percentage:(int)percent {
  RCTLogInfo(@"rn_media_downloader onDownloadingProgress vid = %@-%@ percent = %d", downloader.mm_vid, downloader.mm_trackIndex, percent);

  NSMutableDictionary *body = [NSMutableDictionary dictionary];
  body[@"vid"] = downloader.mm_vid;
  body[@"trackIndex"] = downloader.mm_trackIndex;
  body[@"percent"] = @(percent);
  [self sendEventWithName:EK_downloaderOnDownloadingProgress body:body];
}

/**
 @brief 下载文件的处理进度回调
 @param downloader 下载downloader指针
 @param percent 下载进度 0-100
 */
- (void)onProcessingProgress:(AliMediaDownloader*)downloader percentage:(int)percent {
  RCTLogInfo(@"rn_media_downloader onProcessingProgress vid = %@-%@ percent = %d", downloader.mm_vid, downloader.mm_trackIndex, percent);

  NSMutableDictionary *body = [NSMutableDictionary dictionary];
  body[@"vid"] = downloader.mm_vid;
  body[@"trackIndex"] = downloader.mm_trackIndex;
  body[@"percent"] = @(percent);
  [self sendEventWithName:EK_downloaderOnProcessingProgress body:body];
}

/**
 @brief 下载完成回调
 @param downloader 下载downloader指针
 */
- (void)onCompletion:(AliMediaDownloader*)downloader {
  RCTLogInfo(@"rn_media_downloader onCompletion vid = %@-%@-%@", downloader.mm_vid, downloader.mm_trackIndex, downloader.downloadedFilePath);

  NSMutableDictionary *body = [NSMutableDictionary dictionary];
  body[@"vid"] = downloader.mm_vid;
  body[@"trackIndex"] = downloader.mm_trackIndex;
  body[@"filePath"] = [downloader downloadedFilePath];
  [self sendEventWithName:EK_downloaderOnCompletion body:body];
}
@end

File: android/src/main/java/com/wmmaliyunplayer/RNAliMediaDownloader.java
package com.wmmaliyunplayer;

import android.content.Context;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.aliyun.downloader.AliDownloaderFactory;
import com.aliyun.downloader.AliMediaDownloader;
import com.aliyun.downloader.DownloaderConfig;
import com.aliyun.player.bean.ErrorInfo;
import com.aliyun.player.nativeclass.MediaInfo;
import com.aliyun.player.nativeclass.TrackInfo;
import com.aliyun.player.source.VidSts;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * 第二版，简易版
 * */
public class RNAliMediaDownloader extends ReactContextBaseJavaModule {

    private Context mContext;

    /**
     * AliyunDownloadMediaInfo和AliMediaDownloader 一一 对应
     */
    private LinkedHashMap<AliyunDownloadMediaInfo, AliMediaDownloader> downloadInfos = new LinkedHashMap<>();

    private String downloadDir;
    private DownloaderConfig config;

    @NonNull
    @Override
    public String getName() {
        return "RNAliMediaDownloader";
    }

    public RNAliMediaDownloader(@NonNull ReactApplicationContext reactContext){
        super(reactContext);
        mContext = reactContext;
    }

    @Nullable
    @Override
    public Map<String, Object> getConstants() {
        final Map<String, Object> constants = new HashMap<>();
        constants.put("onPrepared", DownloadEvents.downloaderOnPrepared.toString());
        constants.put("onDownloadingProgress", DownloadEvents.downloaderOnDownloadingProgress.toString());
        constants.put("onError", DownloadEvents.downloaderOnError.toString());
        constants.put("onCompletion", DownloadEvents.downloaderOnCompletion.toString());
        constants.put("onProcessingProgress", DownloadEvents.downloaderOnProcessingProgress.toString());
        return constants;
    }

    /**
     * 准备下载
     * @param vid 视频播放的vid标识
     * @param accessKeyId 用户sts的accessKey ID
     * @param accessKeySecret 用户sts的accessKey secret
     * @param securityToken 用户sts的token信息
     * @param region 用户sts的region信息
     * */
    @ReactMethod
    public void prepareWithVid(final String vid, final String accessKeyId, final String accessKeySecret, final String securityToken, final String region){
        if (TextUtils.isEmpty(vid)){
            return;
        }
        final VidSts vidSts = getVidSts(vid, accessKeyId, accessKeySecret, securityToken, region);
        final AliMediaDownloader jniDownloader = AliDownloaderFactory.create(mContext);
        jniDownloader.setOnPreparedListener(new AliMediaDownloader.OnPreparedListener() {
            @Override
            public void onPrepared(MediaInfo mediaInfo) {
                final WritableMap event = Arguments.createMap();
                event.putString("vid",mediaInfo.getVideoId());
                WritableArray array = Arguments.createArray();
                List<TrackInfo> trackInfos = mediaInfo.getTrackInfos();
                if (trackInfos != null){
                    for (TrackInfo trackInfo : trackInfos) {
                        TrackInfo.Type type = trackInfo.getType();
                        if (type == TrackInfo.Type.TYPE_VOD) {
                            WritableMap map = Arguments.createMap();
                            map.putInt("trackIndex",trackInfo.getIndex());
                            map.putString("trackDefinition",trackInfo.getVodDefinition());
                            map.putString("vodFormat",trackInfo.getVodFormat());
                            map.putInt("vodFileSize", (int) trackInfo.getVodFileSize());
                            array.pushMap(map);

                            final AliyunDownloadMediaInfo downloadMediaInfo = new AliyunDownloadMediaInfo();
                            downloadMediaInfo.setmVid(mediaInfo.getVideoId());
                            downloadMediaInfo.setTrackIndex(trackInfo.getIndex());
                            downloadMediaInfo.setVidSts(vidSts);
                            AliMediaDownloader aliMediaDownloader = downloadInfos.get(downloadMediaInfo);
                            if (aliMediaDownloader == null) {
                                aliMediaDownloader = AliDownloaderFactory.create(mContext);
                            }
                            aliMediaDownloader.updateSource(vidSts);
                            if (!TextUtils.isEmpty(downloadDir)){
                                aliMediaDownloader.setSaveDir(downloadDir);
                            }
                            downloadInfos.put(downloadMediaInfo, aliMediaDownloader);
                        }
                    }
                }
                event.putArray("tracks",array);
                sendEventToRn(DownloadEvents.downloaderOnPrepared.toString(),event);
            }
        });

        AliyunDownloadMediaInfo mediaInfo = new AliyunDownloadMediaInfo();
        mediaInfo.setmVid(vid);
        mediaInfo.setVidSts(vidSts);
        setErrorListener(jniDownloader,mediaInfo);
        if (config != null){
            jniDownloader.setDownloaderConfig(config);
        }
        jniDownloader.prepare(vidSts);

    }

    @ReactMethod
    public void startWithVid(String vid, int trackIndex){
        if (TextUtils.isEmpty(vid)){
            return;
        }
        AliyunDownloadMediaInfo downloadMediaInfo = new AliyunDownloadMediaInfo(vid,trackIndex);
        AliMediaDownloader jniDownloader = downloadInfos.get(downloadMediaInfo);
        if (jniDownloader == null) {
            jniDownloader = AliDownloaderFactory.create(mContext);
            downloadInfos.put(downloadMediaInfo,jniDownloader);
        }
        jniDownloader.start();
    }

    @ReactMethod
    public void stopWithVid(String vid, int trackIndex){
        if (TextUtils.isEmpty(vid)){
            return;
        }
        AliyunDownloadMediaInfo downloadMediaInfo = new AliyunDownloadMediaInfo(vid,trackIndex);
        AliMediaDownloader jniDownloader = downloadInfos.get(downloadMediaInfo);
        if (jniDownloader == null) {
            return;
        }
        jniDownloader.stop();
    }

    @ReactMethod
    public void setSaveDirectory(String dir, String vid, int trackIndex){
        if (TextUtils.isEmpty(dir) || TextUtils.isEmpty(vid)){
            return;
        }
        this.downloadDir = dir;
        AliyunDownloadMediaInfo downloadMediaInfo = new AliyunDownloadMediaInfo(vid,trackIndex);
        AliMediaDownloader jniDownloader = downloadInfos.get(downloadMediaInfo);
        if (jniDownloader == null) {
            jniDownloader = AliDownloaderFactory.create(mContext);
            jniDownloader.setSaveDir(dir);
            downloadInfos.put(downloadMediaInfo,jniDownloader);
        }else {
            jniDownloader.setSaveDir(dir);
        }

    }

    /**
     * 设置下载的trackIndex
     * @param trackIndex 从prepare回调中可以获取所有index
     * */
    @ReactMethod
    public void selectTrack(int trackIndex, String vid){
        if (TextUtils.isEmpty(vid)){
            return;
        }
        AliyunDownloadMediaInfo downloadMediaInfo = new AliyunDownloadMediaInfo(vid,trackIndex);
        AliMediaDownloader jniDownloader = downloadInfos.get(downloadMediaInfo);
        if (jniDownloader == null) {
            jniDownloader = AliDownloaderFactory.create(mContext);
            downloadInfos.put(downloadMediaInfo,jniDownloader);
        }
        jniDownloader.selectItem(trackIndex);
        setListener(downloadMediaInfo, jniDownloader);
    }

    /**
     * 释放资源
     * */
    @ReactMethod
    public void destroyWithVid(String vid, int trackIndex){
        if (TextUtils.isEmpty(vid)){
            return;
        }
        AliyunDownloadMediaInfo info = new AliyunDownloadMediaInfo(vid,trackIndex);
        if(downloadInfos != null && downloadInfos.containsKey(info)){
            AliMediaDownloader aliMediaDownloader = downloadInfos.get(info);
            if(aliMediaDownloader != null){
                aliMediaDownloader.release();
            }
            downloadInfos.remove(info);
        }
    }

    @ReactMethod
    public void updateWithVid(String vid, int trackIndex, String accessKeyId, String accessKeySecret, String securityToken, String region){
        VidSts vidSts = getVidSts(vid, accessKeyId, accessKeySecret, securityToken, region);
        AliyunDownloadMediaInfo info = new AliyunDownloadMediaInfo(vid,trackIndex);
        info.setVidSts(vidSts);
        AliMediaDownloader jniDownloader = downloadInfos.get(info);
        if (jniDownloader == null) {
            jniDownloader = AliDownloaderFactory.create(mContext);
            downloadInfos.put(info,jniDownloader);
        }
        jniDownloader.updateSource(vidSts);
    }

    /**
     * 设置下载配置
     * @param timeoutMs 最大超时时间 默认15000毫秒
     * @param connnectTimoutMs 最大连接超时时间 默认5000毫秒
     * @param referer 请求referer
     * @param userAgent 用户代理
     * @param httpProxy http代理
     * */
    @ReactMethod
    public void setConfig(int timeoutMs, int connnectTimoutMs, String referer, String userAgent, String httpProxy, String vid, int trackIndex){
        if (TextUtils.isEmpty(vid)){
            return;
        }
        config = new DownloaderConfig();
        //连接超时时间
        config.mConnectTimeoutS = connnectTimoutMs > 0? connnectTimoutMs/1000 : 15;
        //网络请求超时时间
        config.mNetworkTimeoutMs = timeoutMs > 0? timeoutMs : 5000;
        if (!TextUtils.isEmpty(referer)){
            config.mReferrer = referer;
        }
        if (!TextUtils.isEmpty(userAgent)){
            config.mUserAgent = userAgent;
        }
        if (!TextUtils.isEmpty(httpProxy)){
            config.mHttpProxy = httpProxy;
        }
        AliyunDownloadMediaInfo downloadMediaInfo = new AliyunDownloadMediaInfo(vid,trackIndex);
        AliMediaDownloader jniDownloader = downloadInfos.get(downloadMediaInfo);
        if (jniDownloader == null) {
            jniDownloader = AliDownloaderFactory.create(mContext);
            jniDownloader.setDownloaderConfig(config);
            downloadInfos.put(downloadMediaInfo,jniDownloader);
        }else {
            jniDownloader.setDownloaderConfig(config);
        }
    }

    /**
     * 删除下载
     * @param saveDir 文件保存路径
     * @param vid 视频播放的vid标识
     * @param format 视频格式
     * @param index 分辨率下标
     * */
    @ReactMethod
    public void deleteFile(String saveDir, String vid, String format, int index, Callback callback){
        if (TextUtils.isEmpty(vid) || TextUtils.isEmpty(saveDir)){
            return;
        }
        int ret = AliDownloaderFactory.deleteFile(saveDir, vid, format, index);
        callback.invoke(ret);
    }


    /**
     * 获取下载之后的文件路径
     * */
    @ReactMethod
    public void downloadedFilePathWithVid(String vid, int trackIndex, Callback callback){
        if (TextUtils.isEmpty(vid)){
            return;
        }
        AliyunDownloadMediaInfo downloadMediaInfo = new AliyunDownloadMediaInfo(vid,trackIndex);
        AliMediaDownloader jniDownloader = downloadInfos.get(downloadMediaInfo);
        if (jniDownloader == null) {
            return;
        }
        String filePath = jniDownloader.getFilePath();
        callback.invoke(filePath);
    }

    /**
     * 构造VidSts实体
     * @param vid   videoId
     */
    private VidSts getVidSts(String vid, String accessKeyId, String accessKeySecret, String securityToken, String region){
        VidSts vidSts = new VidSts();
        vidSts.setVid(vid);
        vidSts.setRegion(region);
        vidSts.setAccessKeyId(accessKeyId);
        vidSts.setSecurityToken(securityToken);
        vidSts.setAccessKeySecret(accessKeySecret);
        return vidSts;
    }

    /**
     * 原生向rn发送的事件定义
     * */
    private enum DownloadEvents {
        //下载准备完成事件回调
        downloaderOnPrepared("downloaderOnPrepared"),
        //错误代理回调
        downloaderOnError("downloaderOnError"),
        //下载进度回调
        downloaderOnDownloadingProgress("downloaderOnDownloadingProgress"),
        //下载文件的处理进度回调
        downloaderOnProcessingProgress("downloaderOnProcessingProgress"),
        //下载完成回调
        downloaderOnCompletion("downloaderOnCompletion");

        private final String mName;

        DownloadEvents(final String name) {
            mName = name;
        }

        @Override
        public String toString() {
            return mName;
        }
    }

    /**
     * 设置监听
     */
    private void setListener(@NonNull final AliyunDownloadMediaInfo downloadMediaInfo, @Nullable final AliMediaDownloader jniDownloader) {
        if (jniDownloader == null){
            return;
        }
        jniDownloader.setOnProgressListener(new AliMediaDownloader.OnProgressListener() {

            @Override
            public void onDownloadingProgress(int percent) {
                if (TextUtils.isEmpty(downloadMediaInfo.getmVid())){
                    return;
                }
                WritableMap event = Arguments.createMap();
                event.putString("vid",downloadMediaInfo.getmVid());
                event.putInt("trackIndex",downloadMediaInfo.getTrackIndex());
                event.putInt("percent",percent);
                sendEventToRn(DownloadEvents.downloaderOnDownloadingProgress.toString(),event);
            }

            @Override
            public void onProcessingProgress(int percent) {
                if (TextUtils.isEmpty(downloadMediaInfo.getmVid())){
                    return;
                }
                WritableMap event = Arguments.createMap();
                event.putString("vid",downloadMediaInfo.getmVid());
                event.putInt("trackIndex",downloadMediaInfo.getTrackIndex());
                event.putInt("percent",percent);
                sendEventToRn(DownloadEvents.downloaderOnProcessingProgress.toString(),event);
            }
        });

        jniDownloader.setOnCompletionListener(new AliMediaDownloader.OnCompletionListener() {
            @Override
            public void onCompletion() {
                if (TextUtils.isEmpty(downloadMediaInfo.getmVid())){
                    return;
                }
                WritableMap event = Arguments.createMap();
                event.putString("vid",downloadMediaInfo.getmVid());
                event.putInt("trackIndex",downloadMediaInfo.getTrackIndex());
                event.putString("filePath",jniDownloader.getFilePath());
                sendEventToRn(DownloadEvents.downloaderOnCompletion.toString(),event);
            }
        });

        setErrorListener(jniDownloader,downloadMediaInfo);
    }
    /**
     * 原生向rn发送消息
     * @param eventName 事件名称
     * @param data 数据
     * */
    private void sendEventToRn(final String eventName, final WritableMap data) {
        ThreadUtils.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                getReactApplicationContext().getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit(eventName,data);
            }
        });
    }

    /**
     * 设置错误监听
     * @param jniDownloader             下载类
     * @param aliyunDownloadMediaInfo   javaBean
     */
    private void setErrorListener(@Nullable final AliMediaDownloader jniDownloader, @NonNull final AliyunDownloadMediaInfo aliyunDownloadMediaInfo) {
        if(jniDownloader == null){
            return ;
        }
        jniDownloader.setOnErrorListener(new AliMediaDownloader.OnErrorListener() {
            @Override
            public void onError(ErrorInfo errorInfo) {
                if (TextUtils.isEmpty(aliyunDownloadMediaInfo.getmVid())){
                    return;
                }
                WritableMap event = Arguments.createMap();
                event.putString("vid",aliyunDownloadMediaInfo.getmVid());
                event.putInt("trackIndex",aliyunDownloadMediaInfo.getTrackIndex());
                event.putString("message",errorInfo.getMsg());
                event.putInt("code",errorInfo.getCode().getValue());
                sendEventToRn(DownloadEvents.downloaderOnError.toString(),event);
            }
        });
    }

}
